src/mbgl/tile/geometry_tile.cpp

```diff
@@ -125,36 +125,17 @@ void GeometryTile::setShowCollisionBoxes(const bool showCollisionBoxes_) {
 }
 
 void GeometryTile::onLayout(LayoutResult result, const uint64_t resultCorrelationID) {
-    // Don't mark ourselves loaded or renderable until the first successful placement
-    // TODO: Ideally we'd render this tile without symbols as long as this tile wasn't
-    //  replacing a tile at a different zoom that _did_ have symbols.
-    (void)resultCorrelationID;
-    nonSymbolBuckets = std::move(result.nonSymbolBuckets);
-    // It is possible for multiple onLayouts to be called before an onPlacement is called, in which
-    // case we will discard previous pending data
-    pendingFeatureIndex = {{ false, std::move(result.featureIndex) }};
-    pendingData = {{ false, std::move(result.tileData) }};
-    observer->onTileChanged(*this);
-}
-
-void GeometryTile::onPlacement(PlacementResult result, const uint64_t resultCorrelationID) {
     loaded = true;
     renderable = true;
     if (resultCorrelationID == correlationID) {
         pending = false;
     }
+    
+    nonSymbolBuckets = std::move(result.nonSymbolBuckets);
     symbolBuckets = std::move(result.symbolBuckets);
-    // When symbolBuckets arrive, mark pendingData/FeatureIndex as "ready for commit" at the
-    // time of the next global placement. We are counting on these symbolBuckets being
-    // in sync with the data/index in the last `onLayout` call, even though the correlation IDs
-    // may not be the same (e.g. "setShowCollisionBoxes" could bump the correlation ID while
-    // waiting for glyph dependencies)
-    if (pendingData) {
-        pendingData->first = true;
-    }
-    if (pendingFeatureIndex) {
-        pendingFeatureIndex->first = true;
-    }
+    
+    dataPendingCommit = {{ std::move(result.tileData), std::move(result.featureIndex) }};
```

Comment:
Since they must be synchronized, can `tileData` be owned by the `FeatureIndex`?

Reply:
:+1: Making this change actually exposed I bug I had previously introduced with this PR. 😬 

Because the `data.clone()` call had moved to symbol layout time, the feature index could be desynchronized from the data if a `setData` arrived while the worker was waiting for symbols (the new `data` is supposed to be queued up for a re-layout after the current one finishes). Moving ownership of the `GeometryTileData` into `FeatureIndex` solves the problem by forcing the clone to happen at the time the index is created.